Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift

```diff
@@ -272,26 +279,20 @@ import BraintreeCore
     private func performV2Authentication(with lookupResult: BTThreeDSecureResult) {
         threeDSecureV2Provider?.process(lookupResult: lookupResult) { result, error in
             guard let result else {
-                self.apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.failed")
+                if let nsError = error as NSError?, nsError.code == BTThreeDSecureError.canceled.errorCode {
+                    self.apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.verifyCanceled)
+                } else {
```

Comment:
Both cancel and error cases get sent as an error from ThreDSecureV2Provider's process function, so instead of sending all the messages separately, I just distinguished the cases here.

Reply:
Should "cancel" from `ThreeDSecureV2Provider` return` challengeFailed`? I think no.  Those are sent from BTThreeDSecureProvider class's `CardinalValidationDelegate` function.

Reply:
I think we need to make MockCardinalSession and change BTThreeDSecureVwProvider to take a protocol, CardinalSessionProtocol. I tried to run unit tests that require cardinal session to complete but tests quit out on continueWith function.

Reply:
I think it's same pattern in PayPal except we deal with ASWebAuthentication cancelLogin.